### PR TITLE
[SYCL] Support sycl::kernel_bundle for multi-device scenario

### PR DESCRIPTION
**Pull request by againull**
This PR includes:
* Changes in the program manager methods to be able to properly create/build UR program for multiple devices. So far, we were mostly using only the first device in the vector to create/build UR program which made UR program unusable on other devices.

* UR tag update brings the version of urProgramCreateWithBinary which allows
  to create UR program from multiple device binaries.

* Our program cache key allowed only a single device. I have changed it to contain a set of devices. If UR program is created and built for a set of devices then the same UR program is usable whenver we have any subset of this set. That's why if we have a program built for a set of devices then add all subsets to the cache. Before we were adding a record to the cache for each device from the set which is incorrect. For example, if someone requests a UR program for {dev2, dev3} from the cache then it is expected that this UR progam must be usable to submit a kernel to dev3. But we could get a program for {dev1, dev2} from the cache which is unusable on dev3.

### CODE REVIEW

**Comment by againull** [comment]
NB. New tests require https://github.com/intel/llvm/pull/15563 to land to pass.

**Review by cperkinsintel — APPROVED** [review]

**Review by pbalcer — COMMENTED** [review]
https://github.com/intel/llvm/pull/15773 needs to be merged before this to pull sanitizer changes that went into ur prior to this.

**Review by nrspruit — APPROVED** [review]
Please merge to fix the build with intel/llvm and latest Unified Runtime mainline.

**Comment by againull** [comment]
Failures:
SYCL :: AOT/double.cpp
SYCL :: AOT/half.cpp
SYCL :: AOT/reqd-sg-size.cpp
are unrelated and going to be fixed by https://github.com/intel/llvm/pull/15823

File: sycl/test-e2e/ProgramManager/multi_device_bundle/device_libs_and_caching.cpp [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+// RUN: %{build} -fsycl-device-lib-jit-link -o %t.out
+
+// Check the default case when in-memory caching is enabled.
+// RUN: env NEOReadDebugKeys=1 CreateMultipleRootDevices=4 SYCL_UR_TRACE=2 %{run} %t.out | FileCheck %s --check-prefixes=CHECK-SPIRV-JIT-LINK-TRACE
+
+// Check the case when in-memory caching of the programs is disabled.
+// RUN: env SYCL_CACHE_IN_MEM=0 NEOReadDebugKeys=1 CreateMultipleRootDevices=4 %{run} %t.out
+
+// Test AOT next.
+// RUN: %{build} -fsycl-targets=spir64_gen -Xsycl-target-backend=spir64_gen "-device *" -o %t.out
```
Comment (AlexeySachkov) [inline sycl/test-e2e/ProgramManager/multi_device_bundle/device_libs_and_caching.cpp:23]: @againull, `ocloc` requirement is missing for this test, making it fail in some environments

Comment (againull) [reply sycl/test-e2e/ProgramManager/multi_device_bundle/device_libs_and_caching.cpp:23]: Thank you, missed that, will fix the test.

**Pull request merged**